fixtures/flight/server/global.js

```diff
@@ -140,10 +141,17 @@ app.all('/', async function (req, res, next) {
       // For HTML, we're a "client" emulator that runs the client code,
       // so we start by consuming the RSC payload. This needs a module
       // map that reverse engineers the client-side path to the SSR path.
-      const root = await createFromNodeStream(rscResponse, moduleMap);
+      let root;
+      let Root = () => {
+        if (root) {
+          return root;
+        }
+        root = createFromNodeStream(rscResponse, moduleMap);
+        return root;
+      };
```

Comment:
This looks like the implementation of what @sebmarkbage mentioned in this [tweet](https://twitter.com/sebmarkbage/status/1643745109987717120?s=20). Can you explain why this change is needed for preloading the client chunks? Does it lead to HTML chunks being emitted earlier?

Reply:
Flight rows are processed outside of a render. For directives coming from Flight like instructions to preload something we still need to be able to associate these outside of render things to a specific request in SSR. By creating the Flight client during a Fizz render we can associate the flight client rows to a Request. If the environment supports AsyncLocalStorage then even async code (really anything outside of a render) can still see this Request. In environments that don’t support asynclocalstorage flight directives will get dropped because there is no associated Fizz request to refer to when not rendering. Currently all directives are optimistic so if we lose some during SSR it’s not going to break anything just be slower. The hope is that we get an AsyncContext or AsyncLocalStorage in every runtime that can do SSR and this will just be the norm. However none of this would work if you just created a random Flight client out of band because the client wouldn’t know where to route anything. We could take other approaches like encoding the directives into the value that will be rendered but that adds a lot of complexity and we think this pattern is sufficient.

Reply:
Thanks a lot for taking the time to write this thorough explanation!

So, we basically need to create the following sequence for preload resources to be associated with the request and thus rendered (I'm just sharing my debugging notes here, in case anyone else wants to follow along):
- `renderToPipeableStream`
- `createRequest` (with resources)
- set current dispatcher to `ReactDOMServerDispatcher`
- `setCurrentResources` (with resources from request) 
- ... \*hand waiving\* render `Root`
- `createFromNodeStream`
- `processFullRow` (case `I`)
- `resolveModule`
- `preloadModule`
- `preloadModuleForSSR`
- `ReactDOMCurrentDispatcher.current.preload` (which is now `ReactDOMServerDispatcher`)
- create preload resource with link chunks and add to `resources.explicitScriptPreloads`
- flush link chunks in `writePreamble`

If `createFromNodeStream` were to be called outside of `Root`, as was the case before in the Flight fixture, the current dispatcher would be (the default?) `ReactDOMClientDispatcher`, and the `preload` call would subsequently fail with `ReferenceError: document is not defined`.

Reply:
There is some nuance here maybe not captured in this linear list but in a handwavy way yeah this is what is happening.

What is important to happen during render is the flight client construction, the actual flight render can be initiated outside of the Fizz render.

Another note is that processFullRow doesn't actually run during the Fizz render even when the flight client is constructed during a Fizz render

Reply:
Makes sense. 👍